Clean up a.d.d. for mpi framework (#607)
pace-neutrons/Horace: documentation/add/04_mpi_framework.md

an information exchange medium, responsible for point-to-point or collective interaction between these processes.

To accommodate different demands of different users on different systems, Horace can utilize three different **message transfer media**, namely:

1. Communications via message files written/read from a storage device,
2. Using Matlab parallel computing toolbox MPI implementation (based on the toolbox lab`**` operations) and finally,
3. Communication using standard MPI framework. Currently supported MPI implementations are Microsoft MPI for Windows and MPICH for Linux.
 
The **communicator/controller** [[1]](https://en.wikipedia.org/wiki/Message_Passing_Interface#Communicator) parts for these media, 
responsible for controlling the parallel processes, are correspondingly:

1. Pool of Matlab sessions or compiled Matlab sessions, launched by Matlab's Java launcher,
2. Matlab Parallel computing toolbox job control mechanism and
3. **mpiexec** launcher, directly controlling standard MPI processes.
 
An additional controller will be necessary to use Horace parallelization on a public cluster. 

> This controller wraps around and interfaces the cluster job submission mechanism (e.g. *qsub* or *bsub*).

Reviewer: `bsub/qsub` are both submission mechanism and a way of querying for state of the job. `slurm` may be mentioned in the row with `bsub/qsub` -- when I was writing the document, I know nothing about `slurm`  A wrapper works both with appropriate submission mechanism and controlling mechanism to submit a job, learn about its progress (together with logging) and quit the job on request.  All these work through the common interface to provide control/information to a user.

Option 1) suits the users who do not want to compile C\++ code and do not have Matlab parallel computing toolbox installed.

Option 2) is best for the people, who has Matlab parallel computing toolbox.

> Option 3) is best for for experienced users who can compile C++ code and set up MPI framework.

Reviewer 2: ```suggestion
Option 3) is best for users who have acess to Herbert's compiled C++ binaries, and do not have access to Matlab's Parallel Computing Toolbox.
```

Realize this wasn't originally written by you, but users shouldn't need to compile anything to run our C++ framework. We package the compiled binaries they need.

To provide simple switching between different frameworks, common wrappers are written around the libraries and programs controlling parallel processes and transferring the messages.
The wrappers provide common interface to user jobs. 

> Horace parallel job tasks are controlled and communicate with each other using this interface, which makes them independent of the underlying communication method.

Author: Not sure. I believe it's 3 wrappers. One around the FileWrapped (Herbert) mechanism, one around MATLAB parallel toolbox (parpool) and one about MPI. @abuts Might be able to clarify.

> Horace parallel job tasks are controlled and communicate with each other using this interface, which makes them independent of the underlying communication method.

Reviewer: indeed -- you select a transfer mechanism (one of 3 supported) and communicate through this mechanism using the common interface.

| `mpiexec MPI`                   | *possible* | ---         | **Native**  |
| `Job submission/initialization` | **Only**   | ---         | ---         |

Where the word *Native* indicates the medium a communicator is designed for (e.g. mpiexec is responsible for controlling the pool of parallel processes/programs, communicating over MPI interface). 
*Possible* means that, despite *Native* mechanisms existing, communication can be performed by alternative means, 
i.e. *mpiexec* MPI processes can communicate between each other sending file-based messages if necessary.
This usually makes sense for debugging purposes only and is described in **ClusterWrapper** classes chapter below.
The last row of the table indicates that each parallel job is initialized and controlled from the *Login Session* (**Fig 1**) using file-based messages mechanism, 
i.e. the initialization information for a parallel task. Logging of the progress and the information about the task results are distributed using file-based messages. 

The File-based framework has been written with the assumption that *job chunks may be processed independently*, 

> but limited number of interprocess communications is occurring. 

Reviewer 2: ```suggestion
and there are limited numbers of interprocess communications. 
```

This restricts the use of file-based framework to the correspondent part of Horace algorithms and for initial job submission. 
Matlab MPI and standard MPI frameworks do not have such restrictions but can not be used for job submission to the cluster.

A cluster, executing a Horace job using the wrappers should have the following simple topology: 

![Fig 1: Horace Cluster](../diagrams/HoraceMPICluster.png)

**Fig 1:** Cluster running Horace job with location of main software components of the parallel framework.

The diagram shows the interaction between the main hardware and software components, used to run these jobs. 
Black boxes in **Fig 1** indicate the hardware components, namely parallel processes or parallel programs running Horace software. 
Green boxes refer to the software components, running on the appropriate hardware. 
The red arrows on the picture refer to the message transfer medium, used for communication between the processes/nodes. 

> Currently implemented Horace jobs mainly communicate with the *head node*, though this is not a constraint, but the an implementation detail. 

Reviewer 2: ```suggestion
Currently implemented Horace jobs mainly communicate with the *head node*, though this is an implementation detail rather than a constraint.
```

> The frameworks allow efficient communication between nodes without any constraints. 

Author: @abuts Any comment on what constraints might be?

> The frameworks allow efficient communication between nodes without any constraints. 

Reviewer: An example of a constrained communication would be an attempt to transfer pixel data using filebased messages within the parallel combine pixel job. This job rearranges pixels to physically place the pixels, belonging to the same cell together, so involves intensive read/write operations. If filebased messages are used to transfer data between nodes, it duplicates the amount of IO operations, necessary to rearrange pixels. Such parallel job is guaranteed to run slower then a serial job. Proper MPI communications, arranged between the nodes (red arrows) are expected to be much faster then IO operations, so may be considered as having minor performance penalty. The communications between the logon session and the head node and, if necessary, logon node and any other worker are still filebased, so are constrained by the performance of filebased operations.  

Hope this explanation clarifies this issue

> The frameworks allow efficient communication between nodes without any constraints. 

Author: So this is a constraint on performance, which still exists on certain frameworks for certain operations?

> The frameworks allow efficient communication between nodes without any constraints. 

Reviewer: generally yes. Filebased messages are relatively inefficient way of communications. They are used for providing initial data for the jobs and one can communicate between the workers and the logon node using them, but in fact its much more efficient to use the MPI channels, directly intended for this purpose.

requiring only one Matlab license for the head node and [Matlab Redistributable](https://uk.mathworks.com/products/compiler/matlab-runtime.html) installed on the cluster. 
A Python wrapper around Horace code, eliminating the need for any Matlab licensing, is under development.

Generic Horace parallel job management and the interaction between software components is presented in **Fig 2**:

![Fig 2: MPI Framework](../diagrams/mpi-framework.png)

**Fig 2:**  Horace parallel job framework

An algorithm with parallel capabilities defines an appropriate **JobExecutor** class, 
which together with correspondent parallel configuration, defined by  **parallel\_config** class is provided to **JobDispatcher**. 
The **JobDispatcher** performs **ClusterWrapper** initialization according to the **parallel_config** 
and the chunking of jobs into parallel "chunks" using a description of the physical problem, provided by **JobExecutor**. 
**ClusterWrapper** starts the parallel cluster and takes the chunks of the job from the **JobDispatcher**, 
formatted as job initialization messages. 
**ClusterWrapper** sends these chunks to the parallel *workers*, which run independent Matlab instances, 

> executing the particular **JobExecutor** job chunks and communicating between each others using the appropriate **MessagesFramework**

Reviewer 2: ```suggestion
executing the particular **JobExecutor** job chunks and communicating between each other using the appropriate **MessagesFramework**
```

The components of a Horace parallel job, presented in **Fig 1** and **Fig 2** and mentioned here are described below. 

(*: one Matlab license for each execution node or Matlab distributed server license with correspondent number of nodes licenses is required)

## Interfaces

A common interface has been implemented to ensure the operation is independent of the underlying service and messaging framework. 

> As we use only a limited subset of MPI operations, necessary to our purposes, the interface is simpler then the standard MPI 
> or even Matlab distributed computing toolbox interface. 

Reviewer 2: ```suggestion
As we use only a limited subset of MPI operations - those necessary to our purpose - 
the interface is simpler than that of standard MPI or Matlab's Parallel Computing Toolbox.
```

| `finalize_all`              | M     | shut down parallel framework and parallel cluster.                                                                                                                                                                                                                                                                                                                                                                                                                     |

A Horace job uses the relevant implementation of  **iMessageFramework** and uses the methods, defined in the interface above to communicate with neighbouring workers when it becomes necessary according to an algorithm's logic. 
Some coarse logic, providing basic communications and synchronization is implemented in the *worker* script.

> [^1]: The developer should expect that error may happen during a worker execution, and a *Fail* or *Canceled* message can be sent from the client instead of regular exchange message. 

Author: Unfortunately, throughout the code it uses "canceled", so this is strictly correct.

If a *receive_message* command is issued in synchronous mode, the frameworks based on the real MPI exchange checks for interrupt once, and enters into synchronous MPI read or wait state, waiting appropriate synchronous message to appear. 
If an interrupt message has been sent instead regular message, indicating error or cancellation issued by other parallel worker, the receivers remain waiting for initial message state. 

> An interript messge usually indicate a critical failure so the framework will be canceled through parallel interrupt, but if this is not happening, the parallel worker will hang up. 

Author: I believe this is hang up, with a corresponding error. A "clean" crash

Currently auxiliary information contains the name of the *message framework* used for communications between workers and worker number (*labNum*) for file-based messages framework. 
To avoid issues with different frameworks and different operating systems processing non-ASCII characters in a command line differently, the location information is encoded into single ASCII-128 string using standard Java base64 encoder. 
The *build_worker_init* method of **iMessagesFramework** performs the encoding of the input initialization information for further transfer of this information to parallel workers using command line arguments of correspondent job.

#### Interrupts or Persistent messages.

Independent workers may need to report to the other workers some special conditions occurring during their execution. 
Such conditions currently result in a failure (an interrupt is raised at code execution). 
We use Persistent messages aka Interrupt messages to distribute information about such states of the system as parallel interrupts, as MPI interrupts used for these purposes in standard MPI frameworks are not available in all parallel frameworks. 

> Such a message, received by a worker from other worker, block all further communication with this worker. 

Reviewer 3: blockS

All requests for information from this worker result in the persistent message (*failure*). 
This state holds until special command (*clear_all* above) is executed by the receiving worker.

#### Messages types.

Different operations can be handled using different types of messages. There are 3 main types, differentiated in Horace.

1. Logging, diagnostics and informing user about progress of a job.
2. Interprocess data exchange, necessary for performing a particular user job and
3. Information about error or failure.

Corresponding types of messages created to serve these purposes:

1. Non-blocking (transient)  messages. Informing user about a job progress is an important task, but often the user is interested only in the final value.
   Different independent tasks should not be synchronous with regard to log messages as only some average progress is required. 
   Non-blocking messages are best suited for such kind of task. 

>    If more then one non-blocking message is present in the MPI messages queue, the client receives only the last message.

Reviewer 3: Does this worry us? Can we improve this?

Reviewer: this is intentional and took a lot of efforts to implement (in some cases)

Author: Why is it intentional? What is the advantage?

Reviewer: if the messages are synchronous, one need to synchronize on them. This synchronization is artificial, as independent processes do not need to know each others progress and synchronization takes time. The purpose of asynchronous messages is mainly logging and in logging you are interested in the last state of an interdependent worker only.

Reviewer 4: If we are trying to trace the execution of a process logging all messages would be useful: the final state will not necessarily contain the information about what triggered a code path that led to failure. 
These are async in as much as we don't need to block the main process execution, but they are still important.
The messages (can) carry information about which process they are from and a timestamp so it doesn't matter if they are interlaced, or even out-of-order in the log files.

3. Blocking messages are necessary to transfer information between independent workers when the result of calculations on a single node depends on the results on other nodes.
4. Persistent messages necessary to indicate some critical conditions or error states, as, for example, 
   parallel interrupts available in MPI are not available in file-based messages system (*ClusterHerbert*) and not accessible for Matlab MPI (*ClusterParpool*). 
   Persistent messages are used to carry such information. 

> To provide the described flexibility and variability of messages, all messages used by Horace MPI are children of **aMessage** class and subscribed to a messages factory (see below). 

Reviewer 3: This is MPI  specific. Wht happens with the other frameworks?

Reviewer: its not MPI specific -- all frameworks communicate using the same messages. Probably word Horace MPI here is confusing -- it points to any parallel framework used by Horace.
